driver-async/src/main/com/mongodb/async/client/gridfs/GridFSBucketImpl.java

```diff
@@ -0,0 +1,401 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.async.client.gridfs;
+
+import com.mongodb.MongoGridFSException;
+import com.mongodb.ReadConcern;
+import com.mongodb.ReadPreference;
+import com.mongodb.WriteConcern;
+import com.mongodb.async.SingleResultCallback;
+import com.mongodb.async.client.MongoClients;
+import com.mongodb.async.client.MongoCollection;
+import com.mongodb.async.client.MongoDatabase;
+import com.mongodb.client.gridfs.model.GridFSDownloadByNameOptions;
+import com.mongodb.client.gridfs.model.GridFSFile;
+import com.mongodb.client.gridfs.model.GridFSUploadOptions;
+import com.mongodb.client.result.DeleteResult;
+import com.mongodb.client.result.UpdateResult;
+import org.bson.BsonDocument;
+import org.bson.BsonObjectId;
+import org.bson.BsonValue;
+import org.bson.Document;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static com.mongodb.internal.async.ErrorHandlingResultCallback.errorHandlingCallback;
+import static java.lang.String.format;
+import static org.bson.codecs.configuration.CodecRegistries.fromRegistries;
+
+@SuppressWarnings("deprecation")
+final class GridFSBucketImpl implements GridFSBucket {
+    private final MongoDatabase database;
+    private final String bucketName;
+    private final int chunkSizeBytes;
+    private final WriteConcern writeConcern;
+    private final ReadConcern readConcern;
+    private final ReadPreference readPreference;
+    private final MongoCollection<Document> filesCollection;
+    private final MongoCollection<Document> chunksCollection;
+    private final CodecRegistry codecRegistry;
+
+    GridFSBucketImpl(final MongoDatabase database) {
+        this(database, "fs");
+    }
+
+    GridFSBucketImpl(final MongoDatabase database, final String bucketName) {
+        this.database = notNull("database", database);
+        this.bucketName = notNull("bucketName", bucketName);
+        this.chunkSizeBytes = 255 * 1024;
+        this.writeConcern = database.getWriteConcern();
+        this.readConcern = database.getReadConcern();
+        this.readPreference = database.getReadPreference();
+        this.codecRegistry = getCodecRegistry();
+        this.filesCollection = getFilesCollection();
+        this.chunksCollection = getChunksCollection();
+    }
+
+    GridFSBucketImpl(final MongoDatabase database, final String bucketName, final int chunkSizeBytes, final CodecRegistry codecRegistry,
+                     final ReadPreference readPreference, final WriteConcern writeConcern, final ReadConcern readConcern,
+                     final MongoCollection<Document> filesCollection, final MongoCollection<Document> chunksCollection) {
```

Comment:
Looks like a bug that filesCollection and chunksCollection are passed as parameters here, as there is no guarantee that they've picked up any new read preference, write concern, read preference, codec registry.

I checked the sync version, and it's the same there.  

To fix it, just don't pass them in and create new ones every time.